cpp/src/join/conditional_join.cu

```diff
@@ -13,31 +13,235 @@
                  ast::expression binary_predicate,
                  null_equality compare_nulls,
                  join_kind JoinKind,
+                 std::optional<std::size_t> output_size,
                  rmm::cuda_stream_view stream,
                  rmm::mr::device_memory_resource* mr)
 {
-  CUDF_FUNC_RANGE();
-  return get_conditional_join_indices(
-    left, right, binary_predicate, compare_nulls, JoinKind, stream, mr);
+  // We can immediately filter out cases where the right table is empty. In
+  // some cases, we return all the rows of the left table with a corresponding
+  // null index for the right table; in others, we return an empty output.
+  if (right.num_rows() == 0) {
+    switch (JoinKind) {
+      // Left, left anti, and full (which are effectively left because we are
+      // guaranteed that left has more rows than right) all return a all the
+      // row indices from left with a corresponding NULL from the right.
+      case join_kind::LEFT_JOIN:
+      case join_kind::LEFT_ANTI_JOIN:
+      case join_kind::FULL_JOIN: return get_trivial_left_join_indices(left, stream);
+      // Inner and left semi joins return empty output because no matches can exist.
+      case join_kind::INNER_JOIN:
+      case join_kind::LEFT_SEMI_JOIN:
+        return std::make_pair(std::make_unique<rmm::device_uvector<size_type>>(0, stream, mr),
+                              std::make_unique<rmm::device_uvector<size_type>>(0, stream, mr));
+    }
+  }
+
+  // Prepare output column. Whether or not the output column is nullable is
+  // determined by whether any of the columns in the input table are nullable.
+  // If none of the input columns actually contain nulls, we can still use the
+  // non-nullable version of the expression evaluation code path for
+  // performance, so we capture that information as well.
+  auto const nullable  = cudf::nullable(left) || cudf::nullable(right);
+  auto const has_nulls = nullable && (cudf::has_nulls(left) || cudf::has_nulls(right));
+
+  auto const parser =
+    ast::detail::expression_parser{binary_predicate, left, right, has_nulls, stream, mr};
+  CUDF_EXPECTS(parser.output_type().id() == type_id::BOOL8,
+               "The expression must produce a boolean output.");
+
+  auto left_table  = table_device_view::create(left, stream);
+  auto right_table = table_device_view::create(right, stream);
+
+  // Allocate storage for the counter used to get the size of the join output
+  detail::grid_1d config(left_table->num_rows(), DEFAULT_JOIN_BLOCK_SIZE);
+  auto const shmem_size_per_block =
+    parser.device_expression_data.shmem_per_thread * config.num_threads_per_block;
+  join_kind KernelJoinKind = JoinKind == join_kind::FULL_JOIN ? join_kind::LEFT_JOIN : JoinKind;
+
+  // If the join size was not provided as an input, compute it here.
+  std::size_t join_size;
+  if (output_size.has_value()) {
+    join_size = *output_size;
+  } else {
+    rmm::device_scalar<std::size_t> size(0, stream, mr);
```

Comment:
I'm not sure I've grasped the significance of having switched from `cudf::size_type` before to `std::size_t` now.

Reply:
The join APIs can return gather maps that have more than `cudf::size_type` rows, which is why they're returning raw RMM buffers rather than `cudf::column` results.  This is touched upon a bit [here](https://github.com/rapidsai/cudf/pull/8928#discussion_r681767936).

Reply:
For context, the goal is to support distributed joins that produce those larger results for which the total size needs to be known a priori.

Reply:
Yep, it's all about avoiding a GPU OOM or cudf `size_type` limit on a column when performing the join.  Knowing the join output size cheaply allows us to batch the join for the cases where we can tell, based on the output row count, that the join maps themselves are going to be too big to manifest directly within our memory budget.  Once the gather maps are manifested, we can subsequently batch the gather of the join input tables to avoid OOM and size limits when manifesting the join results.

Reply:
D'oh! That makes sense.